### ISSUE TITLE
Bug in ICRT Wrapper

### ISSUE THREAD

**Issue by xianjiedai** [Max-Fu/icrt#7]
Dear ICRT team,

Thank you for your excellent work. 

I notice that the latest commit has fixed a problem in the wrapper but there is still one. To reproduce the error, please run the icrt/tools/[inference.ipynb](https://github.com/Max-Fu/icrt/blob/main/tools/inference.ipynb). Details are as the follows:

In "_**Forward pass of ICRT**_", running inference using for loop (with delta_action flag being true):
when the input "action" is not None, the input "action" is of shape [1, action_dim] after _self.prepare_observations()_, while the "proprio" is of shape [1, 1, action_dim]. The additional one dimension corresponds to the "timesteps".

You may want to change the part of code that processes action in _**icrt_wrapper.py**_ (line 191-201), ine 199 specifically. The desired output action shape is probably [Batch_size, timesteps, action_dim]. The missing timestep dim causes the problem.

Could you provide a solution for this? I tried to replace line 199 by "action = torch.tensor(action)[None].float()" but it introduces another problem during running "Prompting" section.

Best regards,
Xianjie

**Comment by Max-Fu** [comment]
Hi Xianjie, 

Can you try replacing line 198 from 
```python
action = convert_delta_action(action[:, None, ...], proprio.numpy().transpose(1, 0, 2)).squeeze()[None, ...]
```
to
```python
action = convert_delta_action(action[:, None, ...], proprio.numpy().transpose(1, 0, 2)).transpose(1, 0, 2)
```
and let me know if that works? 

Best,
Max

**Comment by xianjiedai** [comment]
Hi Max,

Thanks for your quick reply.

It works fine now. I have another question wondering if I could doublecheck it with you.

During training, if the flags:

"**task_barrier**", "**non_overlapping**" and "**sort_by_lang**" set to true, 
"**task_grouping**" and "**rebalance_tasks**" set to false, 

"_self.usable_indices_" and "_find_increasing_subsequences()_" make sure that trajectories within each training batch must come from the same task (verb). However, sicne the "_self.seq_length_" is set to a fix number, the steps retrieved by "__getitem()__" may contain incomplete increasing subsequences, such as the first and last subsequence from [(4, 9), (1,9), (1,3)]. Subsequently, the first subsequence, i.e. (4,9), will be used as prompt.

I am wondering whether it's a better idea to use complete subsequence as prompt instead of incomplete ones. I'd love to hear your comment on this.

Thanks again for your reply and nice work.

Best,
Xianjie

**Comment by xianjiedai** [comment]
Hi again,

I am also wondering whether you could provide the configuration of "Llama-Base", mentioned in the paper as "Llama2 model of 12 layers with a latent dimension of 768". Does the rest of config follows the default setting in ModelArgs (https://github.com/Max-Fu/icrt/blob/main/icrt/models/policy/llama.py ), such as _n_heads_, _max_seq_len_, training epoch, batch size, etc.?

Thanks again and look forward to hearing from you soon!

Best,
Xianjie

**Comment by Max-Fu** [comment]
> I am wondering whether it's a better idea to use complete subsequence as prompt instead of incomplete ones. I'd love to hear your comment on this.

Please do try and I think this may result in better performance. 

> Llama-Base config: 

config/model_config/custom_transformer.json
https://github.com/Max-Fu/icrt/blob/main/config/model_config/custom_transformer.json

Llama2-7B config:
https://huggingface.co/meta-llama/Llama-2-7b/blob/main/params.json

**Comment by xianjiedai** [comment]
Thanks for your reply!

I am closing this issue.

**Comment by xianjiedai** [comment]
Hi Max again,

I have another question regrading the training process and I would love to hear your opinion again.

During training, the image augmentation, including random crop and resize, is applied only on the agent image while the wrist image remains un-augmented. I understand that this setting is coherent with using robot base as the input refenrence frame and probably for the sake of generalization on different environment settings. However, the agent images, within the same trajectory returned by ___getitem___ are randomly cropped and resized. As far as I concern, the _RandomResizedCropAndInterpolation_ from timm is applied on each image independently. I am wondering whether it's better to apply the **same** random resize and crop on images from the same the trajectory, such as using the same crop location and reshape ratio on every image within the same trajectory. 

A naive guess for different random crop on different images within the same batch is that: the cropped image may still include the edges of the table, for example, which can be further used to conjecture the relative position of the obeject to the robot arm, since the arm locates on a fixed position on the table. However, the scale ratio is set to [0.65, 1], which may be a little wide as far as I concern.  Another naive guess would be that the wrist image could provide relative location information of the object as the 3D obejct may look different from different observation angle. This provides the pocily enough relative postion information.

I would like to hear your opinoin on my naive guesses or the motivation of applying random crop on agent images.

I am also curious about other experiment settings. Would random crop/resize be a good idea for the world frame (as input reference frame), such as Libero based on RoboSuite?

Thanks again for your valuable opinion and excellent work.

Best,
Xianjie

**Comment by Max-Fu** [comment]
Hi Xianjie, 

I remember if the RandomResizedCropAndInterpolation is applied on an entire sequence of images, the same randomization is applied. I can look into this again if your visualization suggests otherwise. The reason why we had this randomization was because we were thinking of doing some camera pose randomization experiment, but we never gotten to do it systematically. Otherwise I think it is possible to run it without such randomization. I think you might just have to try it in your setting to see if this is the case.

Best,
Max

**Comment by xianjiedai** [comment]
Thanks for the reply. RandomResizedCropAndInterpolation is applied on the entire sequence simutaneously. Applogize for the misleading.

Best,
Xianjie
